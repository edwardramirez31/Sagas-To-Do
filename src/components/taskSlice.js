/* eslint-disable no-unused-vars */
/* eslint-disable no-param-reassign */
import { createSlice } from '@reduxjs/toolkit';

export const taskSlice = createSlice({
  name: 'task',
  initialState: {
    tasks: [],
    loading: false,
    error: null,
  },
  reducers: {
    addTask: (state, _action) => {
      state.loading = true;
    },
    addTaskSuccess: (state, action) => {
      // { id: 1, text: "Hello", completed: false }
      state.tasks = [...state.tasks, action.payload];
      state.loading = false;
    },
    addTaskError: (state, action) => {
      state.error = action.payload;
      state.loading = false;
    },
    getTask: (state, _action) => {
      state.loading = true;
    },
    getTaskSuccess: (state, action) => {
      // { id: 1, text: "Hello", completed: false }
      state.tasks = [...state.tasks, ...action.payload];
      state.loading = false;
    },
    getTaskError: (state, action) => {
      state.error = action.payload;
      state.loading = false;
    },
    removeTask: (state, action) => {
      state.tasks = state.tasks.filter((task) => task.id !== action.payload);
    },
    setIsUpdating: (state, action) => {
      const newTasks = state.tasks.map((task) =>
        task.id === action.payload ? { ...task, isUpdating: true } : task
      );
      state.tasks = newTasks;
    },
    updateTask: (state, _action) => {
      state.loading = false;
    },
    updateTaskSuccess: (state, action) => {
      const task = action.payload;
      const newTasks = state.tasks.map((item) => {
        if (item.id === task.id) {
          return {
            ...task,
            task: task.text,
            isUpdating: false,
          };
        }
        return item;
      });
      state.tasks = newTasks;
      state.loading = false;
    },
  },
});

export const {
  addTask,
  addTaskSuccess,
  addTaskError,
  getTask,
  getTaskSuccess,
  getTaskError,
  removeTask,
  setIsUpdating,
  updateTask,
  updateTaskSuccess,
} = taskSlice.actions;

export default taskSlice.reducer;
// id, text, completed
// https://edwardramirez.pythonanywhere.com/
// https://edwardramirez.pythonanywhere.com/task/1/
