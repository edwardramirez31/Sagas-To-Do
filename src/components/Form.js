import { Button, TextField } from "@material-ui/core";
import Icon from "@material-ui/core/Icon";
import { Formik } from "formik";
import React from "react";
import { useDispatch } from "react-redux";
import * as Yup from "yup";
import { addTask } from "./taskSlice";

function Form() {
  const dispatch = useDispatch();

  const formSchema = Yup.object().shape({
    text: Yup.string().required(),
  });

  const submitHandler = (data, formikHelpers) => {
    dispatch(addTask({ ...data, completed: false }));
    formikHelpers.resetForm();
  };

  return (
    <Formik
      initialValues={{
        text: "",
      }}
      validationSchema={formSchema}
      onSubmit={submitHandler}
    >
      {({
        values,
        errors,
        touched,
        handleChange,
        handleSubmit,
        handleBlur,
      }) => (
        <form
          style={{
            display: "flex",
            alignItems: "baseline",
            justifyContent: "space-around",
          }}
          onSubmit={handleSubmit}
        >
          <TextField
            label="Add a text"
            style={{ width: "85%" }}
            name="text"
            value={values.text}
            onChange={handleChange}
            error={Boolean(touched.text && errors.text)}
            helperText={touched.text && errors.text}
            onBlur={handleBlur}
          />
          <Button
            type="submit"
            variant="contained"
            color="primary"
            endIcon={<Icon>send</Icon>}
            style={{ width: "10%" }}
          >
            Add
          </Button>
        </form>
      )}
    </Formik>
  );
}

export default Form;
