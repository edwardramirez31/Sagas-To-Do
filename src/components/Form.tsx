import { Button, TextField } from '@material-ui/core';
import Icon from '@material-ui/core/Icon';
import { Formik } from 'formik';
import React from 'react';
import { useDispatch } from 'react-redux';
import * as Yup from 'yup';
import { addTask } from '../app/slices/taskSlice';

const Form: React.VFC = () => {
  const dispatch = useDispatch();

  const formSchema = Yup.object().shape({
    text: Yup.string().required(),
  });

  return (
    <Formik
      initialValues={{
        text: '',
      }}
      validationSchema={formSchema}
      onSubmit={(data, helpers): void => {
        dispatch(addTask({ ...data, completed: false }));
        helpers.resetForm();
      }}
    >
      {({ values, errors, touched, handleChange, handleSubmit, handleBlur }): JSX.Element => (
        <form
          style={{
            display: 'flex',
            alignItems: 'baseline',
            justifyContent: 'space-around',
          }}
          onSubmit={handleSubmit}
        >
          <TextField
            label="Add a text"
            style={{ width: '85%' }}
            name="text"
            value={values.text}
            onChange={handleChange}
            error={Boolean(touched.text && errors.text)}
            helperText={touched.text && errors.text}
            onBlur={handleBlur}
          />
          <Button
            type="submit"
            variant="contained"
            color="primary"
            endIcon={<Icon>send</Icon>}
            style={{ width: '10%' }}
          >
            Add
          </Button>
        </form>
      )}
    </Formik>
  );
};

export default Form;
