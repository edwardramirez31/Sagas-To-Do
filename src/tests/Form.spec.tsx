import { shallow } from 'enzyme';
import { render, screen, waitFor } from '@testing-library/react';
import { useDispatch } from 'react-redux';
import userEvent from '@testing-library/user-event';
import Form from '../components/Form';
import { addTask } from '../app/slices/taskSlice';

jest.mock('react-redux');

describe('Form component', () => {
  const mockedUseDispatch = useDispatch as jest.MockedFunction<typeof useDispatch>;
  const dispatchSpy = jest.fn();
  mockedUseDispatch.mockImplementation(() => dispatchSpy);

  beforeEach(() => {
    mockedUseDispatch.mockClear();
    dispatchSpy.mockClear();
  });

  it('should render as expected', () => {
    const wrapper = shallow(<Form />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should change letters as expected', async () => {
    render(<Form />);
    // initial value should be empty string
    const input = screen.getByDisplayValue('');
    // type something
    userEvent.type(input, 'Mock Task');
    // value typed should be on screen
    expect(screen.getByDisplayValue('Mock Task')).toBeInTheDocument();
    // dispatch an action
    userEvent.click(screen.getByText('Add'));
    // wait for Formik helper to reset values
    await waitFor(() => expect(screen.getByDisplayValue('')).toBeInTheDocument());
    // now value should be empty
    expect(dispatchSpy).toHaveBeenCalledTimes(1);
    // check dispatch call args
    expect(dispatchSpy).toHaveBeenCalledWith(addTask({ text: 'Mock Task', completed: false }));
    expect(screen.queryByDisplayValue('Mock Task')).toBeNull();
  });
});
