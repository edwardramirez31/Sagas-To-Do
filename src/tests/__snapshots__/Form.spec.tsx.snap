// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component should render as expected 1`] = `ShallowWrapper {}`;
